File: API/Helpers/RequestParams/ProductParams.cs
using System.Text;
using API.Helpers.Pagination;

namespace API.Helpers.RequestParams
{
    public class ProductParams : PaginationParams
    {
        public int categoryId { get; set; } = 2;
        public List<int?> SubCategories { get; set; } = new();
        public decimal PriceMin { get; set; } = default;
        public decimal PriceMax { get; set; } = int.MaxValue;
        public int BrandId { get; set; } = default;
        public bool? InStock { get; set; } = null;
        public string SortBy { get; set; }
        public string SortColumn { get; set; }
        public string SearchTerm { get; set; }
        public List<string> Colors { get; set; } = new();

        public override string ToString()
        {
            StringBuilder sb = new();
            foreach (var item in Colors)
            {
                sb.Append(item);
            }
            string colors = sb.ToString();
            sb.Clear();
            foreach (var item in SubCategories)
            {
                sb.Append(item);
            }
            string subCategories = sb.ToString();
            return $"{categoryId}-{subCategories}-{PriceMin}-{PriceMax}-{BrandId}-{InStock}-{SortBy}-{SortColumn}-{SearchTerm}-{colors}";
        }
    }
}

File: client/src/app/_models/Params/ProductParams.ts
// export interface ProductParams {
//   [key: string]: number | number[] | string[];
//   categoryId: number;
//   subCategories: number[];
//   priceStart: number;
//   priceEnd: number;
//   brands: number[];
//   colors: string[];
// }

import { PaginationParams } from "./PaginationParams";

export class ProductParams extends PaginationParams {
  categoryId: number = 2;
  subCategories: number[] = [];
  priceStart: number = 0;
  priceEnd: number = 0;
  searchTerm: string = '';
  inStock: string = '';
  brands: number[] = [];
  colors: string[] = [];
  sortBy: string = 'ascending';
  sortColumn: string = 'rating';
}

export interface FilterParams {
  subCategories: number[];
  priceStart: number;
  priceEnd: number
  brands: number[];
  colors: string[];
}